src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileMetadataService.scala

```diff
@@ -31,35 +38,76 @@ class FileMetadataService(fileMetadataRepository: FileMetadataRepository,
-          _ <- Future(loggingUtils.logFileFormatStatus("checksum", addFileMetadataInput.fileId, fileStatus))
+          fileStatusRow: FilestatusRow = FilestatusRow(uuidSource.uuid, fileMetadataRow.fileid, Checksum, fileStatus, fileMetadataRow.datetime)
+          _ <- Future(loggingUtils.logFileFormatStatus("checksum", fileMetadataRow.fileid, fileStatus))
           row <- fileMetadataRepository.addChecksumMetadata(fileMetadataRow, fileStatusRow)
-        } yield FileMetadata(filePropertyName, row.fileid, row.value)) recover {
-          case e: Throwable =>
-            throw InputDataException(s"Could not find metadata for file ${addFileMetadataInput.fileId}", Some(e))
-        }
-      case _ => Future.failed(InputDataException(s"$filePropertyName found. We are only expecting checksum updates for now"))
+        } yield FileMetadataWithFileId(fileMetadataRow.propertyname, row.fileid, row.value)
+      } recover {
+        case e: Throwable =>
+          throw InputDataException(s"Could not find metadata for file ${fileMetadataRow.fileid}", Some(e))
+      }
+      case _ => Future.failed(InputDataException(s"${fileMetadataRow.propertyname} found. We are only expecting checksum updates for now"))
     }
   }
+
+  def addBulkFileMetadata(addBulkFileMetadataInput: AddBulkFileMetadataInput, userId: UUID): Future[BulkFileMetadata] = {
+    val fileMetadataProperties = addBulkFileMetadataInput.metadataProperties.distinct
+    val timestamp = Timestamp.from(timeSource.now)
+    val uniqueFileIds: Seq[UUID] = addBulkFileMetadataInput.fileIds.distinct
+    for {
+      fileRows <- fileRepository.getAllDescendants(uniqueFileIds)
+      fileIds: Seq[UUID] = fileRows.collect{case fileRow if fileRow.filetype.get == "File" => fileRow.fileid}
+      fileMetadataRows: Seq[FilemetadataRow] = for {
+        fileId <- fileIds
+        fileMetadata <- fileMetadataProperties
+      } yield FilemetadataRow(uuidSource.uuid, fileId, fileMetadata.value, timestamp, userId, fileMetadata.filePropertyName)
+      metadataRows <- fileMetadataRepository.addFileMetadata(fileMetadataRows)
+      fileIdsAndMetadata = metadataRows.foldLeft((Set[UUID](), Set[FileMetadata]())) {
+        (fileIdsAndMetadata, fileMetadataRow) => {
+          (
+            fileIdsAndMetadata._1 + fileMetadataRow.fileid,
+            fileIdsAndMetadata._2 + FileMetadata(fileMetadataRow.propertyname, fileMetadataRow.value)
+          )
+        }
+      }
+    } yield BulkFileMetadata(fileIdsAndMetadata._1.toSeq, fileIdsAndMetadata._2.toSeq)
+  }
+
+  def getFileMetadata(consignmentId: UUID): Future[Map[UUID, FileMetadataValues]] = fileMetadataRepository.getFileMetadata(consignmentId).map {
+    rows =>
+      rows.groupBy(_.fileid).map {
+        case (fileId, fileMetadata) =>
+          val propertyNameMap: Map[String, String] = fileMetadata.groupBy(_.propertyname)
+            .transform((_, value) => value.head.value)
+          fileId -> FileMetadataValues(
```

Comment:
Just a suggestion but we're doing something very similar here https://github.com/nationalarchives/tdr-consignment-api/blob/master/src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala#L104
That method in the FileService doesn't look like it needs to be in the implicit class either. Is it worth extracting the logic out somewhere to save repeating this to think.

Reply:
Yes, I agree although, even though it looks like I added this method, I didn't actually do it; I think it looks that way due to an merge conflict issue, it seems as though @TomJKing created both. @TomJKing, is there a reason why you added an implicit class to the [this](https://github.com/nationalarchives/tdr-consignment-api/blob/master/src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala#L104)?

Reply:
I think I just missed this when I implemented getting the `FileRow` and `FilemetadataRow` with this method in the `FileService` at line 82 (another facepalm moment...)

The implicit class was there to make things a bit easier. 

We should probably be using the `FileService` method and not have two methods doing essentially the same thing? What do you think, the only difference in the return type is that whole `FileRow` is returned instead of just the `FileId`. Whatever is consuming this response should be able to handle this, and do what it needs. Happy to discuss if I'm talking complete nonesense.

Reply:
👍 . Cool, I've pulled the duplicate code out into a method.